### ISSUE TITLE
Ansicolor 0.6 is corrupting input for warnings-ng for MSBuild

### ISSUE THREAD

**Issue by carpnick** [jenkinsci/ansicolor-plugin#154]
AnsiColor plugin is corrupting integration with warnings next generation plugin.  Escape and control codes are being pushed to its input. 
https://github.com/jenkinsci/warnings-ng-plugin

### Test environment

- Jenkins Version: 2.150.3
- AnsiColor Version: 0.6.2
- Hostsystem: Docker

### Expected behavior

Content received by warnings-ng should be clean, previous version full text
```
[2019-03-09T10:56:23.427Z]      2>GenericCodeRunner\CompositeCode.cs(137,32): warning CS1591: Missing XML comment for publicly visible type or member 'CompositeCode.this[int]' [C:\j\4c8033\src\CodeRunner\CodeRunner.csproj]
```

### Actual behavior
Current behavior with 0.6 and above
```
[2019-03-08T19:56:38.336Z] [m[36;1m     2>[m[33;1mGenericCodeRunner\CompositeCode.cs(137,32): warning CS1591: Missing XML comment for publicly visible type or member 'CompositeCode.this[int]' [C:\j\6aa722\src\CodeRunner\CodeRunner.csproj]
```

### Steps to reproduce the behavior

-  Create csproj and solution for C# and visual studio
- Create a public class
- Enable XML file output in the build properties 
- Append this parameter to the MSBuild command line: `/consoleloggerparameters:ForceConsoleColor`

**Comment by carpnick** [comment]
@uhafner FYI.

I originally went down [this](https://github.com/carpnick/analysis-model/commit/b9abbcfb255d9b4914837b3a7fd794ab39842df2) path, but am not sure if the fix is your side or ansicolor side.

**Comment by uhafner** [comment]
I think this is a general problem in the warnings plugin. It makes sense to solve it there (in a similar way you are already proposing).

**Comment by hashar** [comment]
I have encountered a similar issue with the Collapsing Console Sections plugin and it might be related.

The Collapsing Console Sections plugin injects HTML using a `ConsoleAnnotator` like ansicolor.  The extension points are registered with the default ordinal (`0`) and are they are thus processed in a non deterministic order which is set when the extension points are loaded on Jenkins start (I think).

When the ansicolor annotator runs after the collapsing console section one, ansicolor detects the injected HTML and kindly wrap it with a HTML comment (`<!--`, `-->`), which mess up the HTML.

To have the collapsing console sections plugin to act after the build timestamper plugin (it prefixes lines with a timestamp), I have lowered its ordinal that ensure it creates a section that includes the timestamps. As a side effect, the plugin runs after the ansicolor plugin which can no more mess with the injected HTML.

Long description: https://github.com/jenkinsci/collapsing-console-sections-plugin/pull/18


A potential fix for the warnings-ng plugin would be to ensure that the ansicolor ConsoleAnnotator runs with higher precedence. That might be as easy as:
```diff
--- a/src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java
+++ b/src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java
@@ -150,7 +150,7 @@ final class ColorConsoleAnnotator extends ConsoleAnnotator<Object> {
                 : new ColorConsoleAnnotator(colorMapName, nextOpenTags);
     }
 
-    @Extension
+    @Extension(ordinal = 999)
     public static final class Factory extends ConsoleAnnotatorFactory<Object> {
 
         @Override
```

Then `mvn package` and install the resulting `.hpi`.

**Comment by uhafner** [comment]
@hashar Is this a problem with the [warnings plugin](https://github.com/jenkinsci/warnings-ng-plugin) or with the ansi-color plugin?

**Comment by hashar** [comment]
@uhafner if this issue is the same I encountered, that is due to ansicolor.

Specifically in the `ColorConsoleAnnotator.annotate()` method. There is an inner class with an `emitHtml()` method. The code there is messed up whenever another `ConsoleAnnotator` inserted text before ansicolor.

I am not quite sure what `emitHtml()` is doing, so take the following with a grain of salt.  The method apparently tracks a position in the **original text** and compare that with what the AnsiColor is emitting.  And whenever the counters are different, it HTML comments the extra content. Which mean that the warnings  or collapsible selection generated HTML is messed up.

The code originate from a6548c4d955cd5436e9ce995887039dc84e1fbec by @jglick (released with 0.6.0).

**Comment by dblock** [comment]
I suggest someone write a unit test that demonstrates the issue above and we can easily fix that.

**Comment by uhafner** [comment]
I think we can close this issue, since I fixed the handing of color codes quite some time ago in the warnings plugin (it was part of [release 5.1.1](https://github.com/jenkinsci/analysis-model/compare/analysis-model-5.1.0...analysis-model-5.1.1) in March this year).

**Comment by dblock** [comment]
@hashar seems to think otherwise?

**Comment by hashar** [comment]
The warnings plugin got fixed and I had to fix the collapsing section one. I guess the next plugin using the `ConsoleAnnotator` extension point will also be broken by the ansicolor plugin.  It should not comment out at arbitrary position in the string, maybe the commenting can be removed entirely? That might have been done solely for debugging purposes.

**Comment by dwnusbaum** [comment]
> It should not comment out at arbitrary position in the string, maybe the commenting can be removed entirely?

I think that would cause the literal ANSI escape sequences to be visible in the output. Some background: This plugin switched from `ConsoleNote` to `ConsoleAnnotator` in #132 because of some changes to the way logging works for Pipeline jobs (most log-related APIs were moved from being invoked on the master to the agent, though IIRC, that mode in Pipeline is currently disabled because it caused various issues, with no plans to reenable it, so maybe we could revert all of that?). In that PR, Jesse realized there was no way to delete the ANSI escape sequences as had been done with the `ConsoleNote`-based implementation previously, so instead, he had to [comment out the escape sequences](https://github.com/jenkinsci/ansicolor-plugin/pull/132#discussion_r230946050) (changed from `display: none` to `<!-- -->` in #142) using an algorithm that tries to detect the positions of the escape sequences based on comparing how many characters were written before and after stripping the escape sequences. See #143 and #147 for some other fun issues with this algorithm that have been fixed, and #151 for a class of issues that is still broken.

Like you pointed out, this doesn't work if other `ConsoleAnnotators` modify the same text, since the algorithm will be confused about what text should be commented out. I think your proposal of adding `@Extension(ordinal = 999)` to the `ConsoleAnnotator` here is probably the easiest solution for now.

**Comment by jglick** [comment]
> it caused various issues, with no plans to reenable it

Getting harder to recall details now, but I think everything important was fixed, and the performance was better than the original mode, so it would still be desirable to roll that forward. Anyway probably off topic here.

The `ordinal` patch may be good enough, if it demonstrably solves issues. Most other CI systems (including Blue Ocean) have ANSI support baked in.
